istioctl/cmd/kubeutils.go

```diff
@@ -0,0 +1,169 @@
+// Copyright Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+
+	"istio.io/pkg/env"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+// k8sClientConfig first tries to get the config from inside a kubernetes environment
+// If unsuccessful, it tries to get the config from the default kubeconfig file
+func k8sClientConfig() (*rest.Config, error) {
+	config, err := rest.InClusterConfig()
+	var ocErr error
+	if err != nil {
+		config, ocErr = clientcmd.BuildConfigFromFlags("", kubeconfigPath())
+		if ocErr != nil {
+			return nil, fmt.Errorf("in-cluster error: %v, out-of-cluster error: %v", err, ocErr)
+		}
+	}
+	return config, nil
+}
+
+// Returns a kubernetes clientset with the default config
+func k8sClientset() (*kubernetes.Clientset, error) {
```

Comment:
We should never have multiple k8s clients. At start of the command, we should initialize it, then use that throughout. This will also resolve some of my other comments about duplication. See `pkg/kube`